lib/SPIRV/SPIRVBuiltinHelper.cpp

```diff
@@ -211,3 +210,71 @@ BuiltinCallMutator BuiltinCallHelper::mutateCallInst(CallInst *CI,
                                                      std::string FuncName) {
   return BuiltinCallMutator(CI, std::move(FuncName), Rules, NameMapFn);
 }
+
+Value *BuiltinCallHelper::addSPIRVCall(IRBuilder<> &Builder, spv::Op Opcode,
+                                       Type *ReturnTy, ArrayRef<Value *> Args,
+                                       ArrayRef<Type *> ArgTys,
+                                       const Twine &Name) {
+  // Sanitize the return type, in case it's a TypedPointerType.
+  if (auto *TPT = dyn_cast<TypedPointerType>(ReturnTy))
+    ReturnTy = PointerType::get(TPT->getElementType(), TPT->getAddressSpace());
+
+  // Copy the types into the mangling info.
+  BuiltinFuncMangleInfo BtnInfo;
+  for (unsigned I = 0; I < ArgTys.size(); I++) {
+    if (Args[I]->getType()->isPointerTy()) {
+      assert(cast<PointerType>(Args[I]->getType())
+                 ->isOpaqueOrPointeeTypeMatches(
+                     cast<TypedPointerType>(ArgTys[I])->getElementType()));
+      BtnInfo.getTypeMangleInfo(I).PointerTy = ArgTys[I];
+    }
+  }
+
+  // Create the function and the call.
+  auto *F = getOrCreateFunction(M, ReturnTy, getTypes(Args),
+                                getSPIRVFuncName(Opcode), &BtnInfo);
+  return Builder.CreateCall(F, Args, ReturnTy->isVoidTy() ? "" : Name);
+}
+
+Type *BuiltinCallHelper::adjustImageType(Type *T, StringRef OldImageKind,
+                                         StringRef NewImageKind) {
+  if (auto *TypedPtrTy = dyn_cast<TypedPointerType>(T)) {
+    Type *StructTy = TypedPtrTy->getElementType();
+    // Adapt opencl.* struct type names to spirv.* struct type names.
+    if (isOCLImageType(T)) {
+      auto ImageTypeName = StructTy->getStructName();
+      StringRef Acc = kAccessQualName::ReadOnly;
+      if (hasAccessQualifiedName(ImageTypeName))
+        Acc = getAccessQualifierFullName(ImageTypeName);
+      StructTy = getOrCreateOpaqueStructType(
+          M, mapOCLTypeNameToSPIRV(ImageTypeName, Acc));
+    }
+
+    // Change type name (e.g., spirv.Image -> spirv.SampledImg) if necessary.
+    StringRef Postfixes;
+    if (isSPIRVStructType(StructTy, OldImageKind, &Postfixes))
+      StructTy = getOrCreateOpaqueStructType(
+          M, getSPIRVTypeName(NewImageKind, Postfixes));
+    else {
+      report_fatal_error("Type did not have expected image kind");
+    }
+    return TypedPointerType::get(StructTy, TypedPtrTy->getAddressSpace());
+  }
+  report_fatal_error("Expected type to be a SPIRV image type");
```

Comment:
nit: May be worth to move to an early exit, like:
```
if (!isa<TypedPointerType>(T))
  report_fatal_error("Expected type to be a SPIRV image type");
if (auto *TypedPtrTy = cast<TypedPointerType>(T)) {
}
```

Reply:
I intend for there to be a future
```c++
if (auto *OpaqueTy = dyn_cast<OpaqueType>(T)) {
  // Whee, do more stuff!
}
```
if statement, so I'm averse to the early exit here.

Reply:
Got it, thanks!